cpp/daal/src/algorithms/dtrees/gbt/regression/gbt_regression_predict_dense_default_batch_impl.i

```diff
@@ -220,26 +291,320 @@ services::Status PredictRegressionTask<algorithmFPType, cpu>::run(const gbt::reg
+        {
+            DAAL_ASSERT(uniquePath[i].partialWeight == 0);
+        }
+    }
+
+    // for (int i = uniqueDepth - 1; i >= 0; --i)
+    // {
+    //     if (oneFraction != 0)
+    //     {
+    //         const float tmp = nextOnePortion * (uniqueDepth + 1) / static_cast<float>((i + 1) * oneFraction);
+    //         total += tmp;
+    //         nextOnePortion = uniquePath[i].partialWeight - tmp * zeroFraction * ((uniqueDepth - i) / static_cast<float>(uniqueDepth + 1));
+    //     }
+    //     else if (zeroFraction != 0)
+    //     {
+    //         total += (uniquePath[i].partialWeight / zeroFraction) / ((uniqueDepth - i) / static_cast<float>(uniqueDepth + 1));
+    //     }
+    //     else
+    //     {
+    //         DAAL_ASSERT(uniquePath[i].partialWeight == 0);
+    //     }
+    // }
+    return total;
+}
+
+/**
+ * Helper to predict SHAP contribution values
+ * \param[in] iTree index of start tree for the calculation
+ * \param[in] nTrees number of trees in block included in calculation
+ * \param[in] nRowsData number of rows to process
+ * \param[in] nColumnsData number of columns in data, i.e. features
+ *                         note: 1 SHAP value per feature + bias term
+ * \param[in] x pointer to the start of observation data
+ * \param[out] res pointer to the start of memory where results are written to
+ * \param[in] dim DimType helper
+*/
+template <typename algorithmFPType, CpuType cpu>
+template <bool hasUnorderedFeatures, bool hasAnyMissing>
+void PredictRegressionTask<algorithmFPType, cpu>::predictContributions(size_t iTree, size_t nTrees, size_t nRowsData, size_t nColumnsData,
+                                                                       const algorithmFPType * x, algorithmFPType * res, const DimType & dim)
+{
+    // TODO: Make use of vectorBlockSize, similar to predictByTreesVector
+
+    const size_t nColumnsPhi   = nColumnsData + 1;
+    const size_t biasTermIndex = nColumnsPhi - 1;
+
+    for (size_t iRow = 0; iRow < nRowsData; ++iRow)
+    {
+        const algorithmFPType * currentX = x + (iRow * nColumnsData);
+        algorithmFPType * phi            = res + (iRow * nColumnsPhi);
+        for (size_t currentTree = iTree; currentTree < iTree + nTrees; ++currentTree)
+        {
+            // regression model builder tree 0 contains only the base_score and must be skipped
+            if (currentTree == 0) continue;
+
+            // prepare memory for unique path data
+            const int depth = _aTree[currentTree]->getMaxLvl() + 2;
+            std::vector<PathElement> uniquePathData((depth * (depth + 1)) / 2);
+
+            // TODO: Not using a separate variable (test) for the phi values causes a lot of cache misses and the code
+            //       runs 10x slower - why?
+            std::vector<algorithmFPType> test(nColumnsData, 0);
+            treeShap<hasUnorderedFeatures, hasAnyMissing>(currentTree, currentX, test.data(), nColumnsData, 1, 0, 0, uniquePathData.data(), 1, 1, -1,
```

Comment:
Too many parameters. What about passing them as structures?

Reply:
I've simplified the user-facing function. The internal (recursive) part is keeping its many params. I don't want to deviate too much from other TreeSHAP implementations.